linker-private-library/src/main/scala/org/scalajs/linker/runtime/RuntimeLong.scala

```diff
@@ -1087,77 +1092,392 @@ object RuntimeLong {
-     * we let the optimizer inline it, it will first store remHi in a temporary
-     * val, which will explose the while condition as a while(true) + if +
-     * break, and we don't want that.
-     */
-    while (shift >= 0 && (remHi & SafeDoubleHiMask) != 0) {
-      if (inlineUnsigned_>=(remLo, remHi, bShiftLo, bShiftHi)) {
-        val newRem =
-          new RuntimeLong(remLo, remHi) - new RuntimeLong(bShiftLo, bShiftHi)
-        remLo = newRem.lo
-        remHi = newRem.hi
-        if (shift < 32)
-          quotLo |= (1 << shift)
-        else
-          quotHi |= (1 << shift) // == (1 << (shift - 32))
+    // scalastyle:off return
+
+    val result = if (bothZero(bhi, blo & 0xffe00000)) {
+      // b < 2^21
+
+      if (askQuotient) {
+        val quotHi = Integer.divideUnsigned(ahi, blo)
+        val k = ahi - blo * quotHi
+        val quotLo = rawToInt(asSafeDouble(alo, k) / blo.toDouble)
+        new RuntimeLong(quotLo, quotHi)
+      } else {
+        val k = Integer.remainderUnsigned(ahi, blo)
+        val quotLo = rawToInt(asSafeDouble(alo, k) / blo.toDouble)
+        val remLo = alo - blo * quotLo
+        new RuntimeLong(remLo, 0)
       }
-      shift -= 1
-      val newBShift = new RuntimeLong(bShiftLo, bShiftHi) >>> 1
-      bShiftLo = newBShift.lo
-      bShiftHi = newBShift.hi
+    } else if ((bhi & 0xc0000000) == 0) {
+      // 2^21 <= b < 2^62
+
+      val a = new RuntimeLong(alo, ahi)
+      val b = new RuntimeLong(blo, bhi)
+      val aHat = unsignedToDoubleApprox(alo, ahi)
+      val bHat = unsignedToDoubleApprox(blo, bhi)
+      val qHat = fromUnsignedSafeDouble(aHat / bHat)
+      val rHat = a - b * qHat
+
+      if (rHat.hi < 0) {
+        if (askQuotient) qHat - new RuntimeLong(1, 0)
+        else rHat + b
+      } else if (geu(rHat, b)) {
+        if (askQuotient) qHat + new RuntimeLong(1, 0)
+        else rHat - b
+      } else {
+        if (askQuotient) qHat
+        else rHat
+      }
+    } else {
+      // 2^62 <= b
+      return unsignedDivModHugeDivisor(alo, ahi, blo, bhi, askQuotient)
     }
 
-    // Now rem < 2^53, we can finish with a double division
-    if (inlineUnsigned_>=(remLo, remHi, blo, bhi)) {
-      val remDouble = asSafeDouble(remLo, remHi)
-      val bDouble = asSafeDouble(blo, bhi)
+    hiReturn = result.hi
+    result.lo
 
-      if (askQuotient) {
-        val rem_div_bDouble = fromUnsignedSafeDouble(remDouble / bDouble)
-        val newQuot = new RuntimeLong(quotLo, quotHi) + rem_div_bDouble
-        hiReturn = newQuot.hi
-        newQuot.lo
+    // scalastyle:on return
+  }
+
+  private def unsignedDivModHugeDivisor(alo: Int, ahi: Int, blo: Int, bhi: Int,
+      askQuotient: Boolean): Int = {
+
+    /* Called for b >= 2^62.
+     * Such huge divisors are practically useless, but they defeat the
+     * correction code of the algorithm above.
+     *
+     * Since b >= 2^62 and a < 2^64, we know that a < 4*b (mathematically).
```

Comment:
Fixed `a <= 2^64` into `a < 2^64`. (If it could be `== 2^64`, the conclusion would be wrong.)